### PR TITLE
Add network switching capabilities to Truffle Dashboard

### PR DESCRIPTION
**Pull request by MicaiahReid**

### CODE REVIEW

**Comment by haltman-at** [comment]
Hm, CI isn't running for some reason...

File: packages/dashboard-message-bus/lib/message/types.ts [hunk 1]
```diff
@@ -31,6 +31,16 @@ export interface LogMessage extends Message {
   };
 }
 
+/**
+ * Message to log in Dashboard browser console
+ */
+export interface DebugMessage extends Message {
+  type: "debug";
```
Comment (cds-amal) [inline packages/dashboard-message-bus/lib/message/types.ts:38]: Perhaps it's time to refine the Message interface to keep track of valid types. @gnidan ?

```typescript

type MessageType = "debug" | "provider" | "log" // | etc...
export interface Message {
  id: number;
  type: MessageType;
  payload: any;
}
```

File: packages/dashboard/lib/DashboardServer.ts [hunk 2]
```diff
@@ -76,8 +101,11 @@ export class DashboardServer {
 
     this.expressApp.get("/ports", this.getPorts.bind(this));
 
+    this.subscribeSocket = await this.connectToSubscribePort();
+
     if (this.rpc) {
```
Comment (MicaiahReid) [inline packages/dashboard/lib/DashboardServer.ts:106]: This probably is legacy code that should be removed. In what circumstances would a user want RPC disabled here?

File: packages/dashboard/src/Dashboard.tsx [hunk 3]
```diff
@@ -19,36 +23,40 @@ import ConfirmNetworkChanged from "./components/ConfirmNetworkChange";
...
     } else {
       setConnectedChainId(chainId);
     }
-  }, [chainId, connectedChainId, socket]);
+  }, [chainId, connectedChainId, subscribeSocket]);
 
-  const initializeSocket = async () => {
-    if (socket && socket.readyState === WebSocket.OPEN) return;
+  const initializeSockets = async () => {
+    await initializeSubSocket();
```
Comment (MicaiahReid) [inline packages/dashboard/src/Dashboard.tsx:49]: Put these in a Promise.all so they can happen RACE.

File: packages/dashboard/src/Dashboard.tsx [hunk 4]
```diff
@@ -70,31 +78,82 @@ function Dashboard() {
...
+
+  const initializePubSocket = async () => {
+    if (publishSocket && publishSocket.readyState === WebSocket.OPEN) return;
 
-    connectedSocket.send("ready");
+    const { publishPort } = await getPorts();
+    const socket = await initializeSocket(
+      publishPort,
+      (event: WebSocket.MessageEvent) => {
+        if (typeof event.data !== "string") {
```
Comment (MicaiahReid) [inline packages/dashboard/src/Dashboard.tsx:97]: The publish socket probably doesn't need any event handler at this point. We can use the socket to make a request for data but we don't need every message.

File: packages/dashboard/src/Dashboard.tsx [hunk 5]
```diff
@@ -70,31 +78,82 @@ function Dashboard() {
...
+              JSON.stringify(logMessage.payload.message)
+          );
+        }
+      }
+    );
+
+    // since our socket is open, request some initial data from the server
+    const message = createMessage("initialize", ""); // no payload needed
+    const response = await sendAndAwait(socket, message);
+    if (response.payload.error) {
```
Comment (MicaiahReid) [inline packages/dashboard/src/Dashboard.tsx:116]: Verify this is actually how the error would show up.

File: packages/dashboard/lib/DashboardServer.ts (hunk 2 above)
Comment (MicaiahReid) [reply packages/dashboard/lib/DashboardServer.ts:106]: Removed.

**Comment by gnidan** [comment]
Hm there's a `package-lock.json` that shouldn't be here

**Comment by MicaiahReid** [comment]
Closing in favor of #5045

**Pull request closed without merging**